tempesta_fw/http.c

```diff
@@ -2916,125 +2949,145 @@ tfw_h2_req_set_loc_hdrs(TfwHttpReq *req)
-				val_found = true;			\
-			} else if (*c->data == ':') {			\
-				/* For statically configured adjustments. */\
-				WARN_ON_ONCE(c->len != SLEN(S_DLM));	\
-				WRITE_LIT(buf, S_DLM);			\
-				val_found = true;			\
-				continue;				\
-			}						\
-		}							\
-		memcpy_fast(buf, c->data, c->len);			\
-		buf += c->len;						\
-	}								\
-	WRITE_LIT(buf, S_CRLF);						\
-})
-
-	T_DBG3("%s: it->hdrs_len=%lu, it->hdrs_cnt=%u\n", __func__,
-	       it->hdrs_len, it->hdrs_cnt);
-
-	/* Substitution/addition of 'x-forwarded-for' header. */
-	hid = __h2_hdr_lookup(hm, TFW_STR_CHUNK(&h_xff, 0));
-	if (unlikely(hid == ht->size))
-		if (tfw_http_msg_grow_hdr_tbl(hm))
-			return -ENOMEM;
-	if (hid < ht->off)
-		__h2_hdrs_dup_decrease(req, &ht->tbl[hid]);
-	else
-		++ht->off;
+	T_DBG3("%s: req [%p] to be converted to http1.1\n", __func__, req);
 
-	++it->hdrs_cnt;
-	it->hdrs_len += h_xff.len;
-	ht->tbl[hid] = h_xff;
+	/* H2 client may use either authority or host header but at least one
+	 * is sufficient for correct conversion.
+	 */
+	if (WARN_ON_ONCE(!auth && !host))
+		return -EINVAL;
 
 	/*
-	 * Conditional substitution/addition/deletion or appending of statically
-	 * configured headers.
+	 * First apply message modifications defined by admin in configuration
+	 * file. Ideally we should do it al last stage, when h2 headers are
+	 * copied into h1 buffer and apply modifications during copying. But
+	 * this doesn't allow us to predict h1 headers size before memory
+	 * allocation. Header modifications manual on wiki is already has a
+	 * warning about performance impact, so just live it as is, a more
+	 * robust algorithm will be used here if really required.
 	 */
 	if ((r = tfw_h2_req_set_loc_hdrs(req)))
 		return r;
 
+	/*
+	 * Calculate http1.1 headers size. H2 request contains pseudo headers
+	 * that are represented in different way in the http1.1 requests.
+	 * pit->hdrs_cnt is aware of header duplicates. Redirection mark is
+	 * ignored and not copied.
+	 */
+	h1_hdrs_sz = pit->hdrs_len
+		+ (pit->hdrs_cnt - pseudo_num) * (SLEN(S_DLM) + SLEN(S_CRLF))
+		- req->mark.len;
+	/* First request line: remove pseudo headers names, all values are on
+	 * the same line.
+	 */
+	h1_hdrs_sz += (long int)2 + SLEN(S_VERSION11) + SLEN(S_CRLF)
+			- ht->tbl[TFW_HTTP_HDR_H2_SCHEME].len
+			- SLEN(S_H2_METHOD)
+			- SLEN(S_H2_PATH)
+			+ SLEN(S_CRLF) /* After headers */;
+	/* :authority pseudo header */
+	if (auth) {
+		/* RFC 7540:
+		 * An intermediary that converts an HTTP/2 request to HTTP/1.1
+		 * MUST create a Host header field if one is not present in a
+		 * request by copying the value of the :authority pseudo-header
+		 * field.
+		 * AND
+		 * Clients that generate HTTP/2 requests directly SHOULD use
+		 * the :authority pseudo-header field instead of the Host
+		 * header field.
+		 */
+		if (host) {
+			h1_hdrs_sz -= ht->tbl[TFW_HTTP_HDR_HOST].len;
+			h1_hdrs_sz -= SLEN(S_H2_AUTH);
+			h1_hdrs_sz += SLEN(S_F_HOST);
```

Comment:
Looks like excessive `SLEN(S_DLM)` counted here for `Host` header; it seems, that this size has already counted above, in the following code:
```
h1_hdrs_sz = pit->hdrs_len
		+ (pit->hdrs_cnt - pseudo_num) * (SLEN(S_DLM) + SLEN(S_CRLF))
		- req->mark.len;
```

Reply:
Fixed, please check once more time.

Reply:
I mean - from the previous code:
```
if (host) {
                        h1_hdrs_sz -= ht->tbl[TFW_HTTP_HDR_HOST].len;
                        h1_hdrs_sz -= SLEN(S_H2_AUTH);
                        h1_hdrs_sz += SLEN(S_F_HOST);
                }
                else {
                        h1_hdrs_sz -= SLEN(S_H2_AUTH);
                        h1_hdrs_sz += SLEN(S_F_HOST) + SLEN(S_CRLF);
                }
```
just to make 
```
if (host) {
                        h1_hdrs_sz -= ht->tbl[TFW_HTTP_HDR_HOST].len;
                        h1_hdrs_sz -= SLEN(S_H2_AUTH);
                        h1_hdrs_sz += SLEN(S_F_HOST) - SLEN(S_DLM);
                }
                else {
                        h1_hdrs_sz -= SLEN(S_H2_AUTH);
                        h1_hdrs_sz += SLEN(S_F_HOST) + SLEN(S_CRLF);
                }

```
to evict `SLEN(S_DLM)` length from `SLEN(S_F_HOST)`, since `SLEN(S_DLM)` for `Host` header (as well as `SLEN(S_CRLF)`) is already contained in `h1_hdrs_sz` in case of `Host` header existence in source HTTP/2-request.

Reply:
Let me clear up the calculations:
- `:authority` psedo-header _is counted_ in `h1_hdrs_sz`,
- if `host` header is present it is counted in `h1_hdrs_sz` and `SLEN(S_DLM) + SLEN(S_CRLF)` was added.

So here we need to make some alterations (assuming that the `:authority` pseudo header exists):

If the `host` header exists, we replace its value with the `:authority` ones. So we  need to:
- subtract  original length of `host` header and length of `S_DLM` and `S_CRLF` added to it;
- we need subtract difference between `:authority` and `host` length: 
```
h1_hdrs_sz -= SLEN(S_H2_AUTH);
/* S_F_HOST already contains S_DLM, but we already calculated `S_DLM` for authority header. */
h1_hdrs_sz += SLEN(S_F_HOST) - SLEN(S_DLM);
```

If host header doesn't exist, the we only need to rename `:authority` into `host`.

That's why che calculations are done in current way.

Reply:
>:authority psedo-header is counted in h1_hdrs_sz

Yes, but in previous version of code the `:authority` pseudo-header had not been counted in `pit->hdrs_cnt` (i.e. subtracted from it), so the `SLEN(S_DLM) + SLEN(S_CRLF)` was not added for it and  - that's why my previous comment appeared. But with last correction in ['fix header calculations'](https://github.com/tempesta-tech/tempesta/pull/1374/commits/b4e2719509d7afd010fba99ae7556781d1c0b056) commit - it seems the problem is gone.
By the way: the `pseudo_num` local variable is constant now and used only once, thus maybe it should be better to move its definition to the beginning of the function, or just evict it (and replace with `3` in the place of usage with adding appropriate comment there).